python/hail/expr/types.py

```diff
@@ -4,23 +4,73 @@
 from hail.history import *
 from hail.typecheck import *
 from hail.utils import Struct
-from hail.utils.java import scala_object, jset, jindexed_seq, Env
+from hail.utils.java import scala_object, jset, jindexed_seq, Env, escape_parsable
 from hail import genetics
+from hail.expr.type_parsing import type_grammar, type_node_visitor
 
 
-class TypeCheckError(Exception):
-    """
-    Error thrown at mismatch between expected and supplied python types.
+def dtype(type_str):
+    r"""Parse a type from its string representation.
 
-    :param str message: Error message
-    """
+    Examples
+    --------
+    .. doctest::
 
-    def __init__(self, message):
-        self.msg = message
-        super(TypeCheckError).__init__(TypeCheckError)
+        >>> hl.dtype('int')
+        dtype('int32')
 
-    def __str__(self):
-        return self.msg
+        >>> hl.dtype('float')
+        dtype('float64')
+
+        >>> hl.dtype('array<int32>')
+        dtype('array<int32>')
+
+        >>> hl.dtype('dict<str, bool>')
+        dtype('dict<str, bool>')
+
+        >>> hl.dtype('struct{a: int32, `field with spaces`: int64}')
+        dtype('struct{a: int32, `field with spaces`: int64}')
+
+    Notes
+    -----
+    This function is able to reverse ``str(t)`` on a :class:`.Type`.
+
+    The grammar is defined as follows:
+
+    .. code-block:: text
+
+        type = array / set / dict / struct / interval / int64 / int32 / float32 / float64 / bool / str / call / str / locus
+        int64 = "int64" / "tint64"
+        int32 = "int32" / "tint32" / "int" / "tint"
+        float32 = "float32" / "tfloat32"
+        float64 = "float64" / "tfloat64" / "tfloat" / "float"
+        bool = "tbool" / "bool"
+        call = "tcall" / "call"
+        str = "tstr" / "str"
+        locus = ("tlocus" / "locus") _ "[" _ identifier _ "]"
+        array = ("tarray" / "array") _ "<" _ type _ ">"
+        set = ("tset" / "set") _ "<" _ type _ ">"
+        dict = ("tdict" / "dict") _ "<" type  _ "," _ type ">"
+        struct = ("tstruct" / "struct") _ "{" _ fields? _ "}"
+        fields = field _ ("," _ field)*
+        field = identifier _ ":" _ type
+        interval = ("tinterval" / "interval") _ "<" _ type _ ">"
+        identifier = simple_identifier / escaped_identifier
+        simple_identifier = ~"\w+"
+        escaped_identifier = "`" ~"([^`\\\\]|\\\\.)*" "`"
+        _ = ~"\s*"
```

Comment:
Since the actual parsimonious grammar definition is just a string, you could use the actual grammar string in the docstring here, to ensure they always agree.

Reply:
I'll try that!

Reply:
I can't get this to work. any idea how?

Reply:
I was just thinking in type_parsing.py you could define a `type_grammar_string = r"""..."""`, then `type_grammar = Grammar(type_grammar_string)`, and in the docstring here you could use `format` to insert `type_grammar_string`. Though now that I think about it, there's probably some indentation issues to handle.

Reply:
Python didn't like formatting the docstring -- it stopped treating it as the docstring and started treating it as an expression.

Reply:
Oh I see. Maybe this isn't possible then. That's unfortunate.